<a href="https://colab.research.google.com/github/preetis258/my_first_repo/blob/master/credit%20card%20fraud%20detection%20using%20Perceptron%20.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Perceptron

In [91]:
import numpy as np

class Perceptron:
  def __init__(self,learning_rate, n_iters, function):
    self.lr = learning_rate #learning rate
    self.n_iters = n_iters #number of iterations
    self.weight = None #weights
    self.bias = None #bias
    if function == 'step_function':
      self.activation_func = self._unit_step_func       #taking step function as activation function
    elif function =='sigmoid':
      self.activation_func = self._sigmoid_func       #taking sigmoid function as activation function
    elif function == 'relu':
      self.activation_func = self._relu_func      #taking relu function as activation function
    elif function == 'tanh':
      self.activation_func = self._tanh_func      #taking tanh function as activation function      

  def _unit_step_func(self,x):
    return np.where(x>=0, 1, 0)

  def _sigmoid_func(self,x):
    return 1/(1 + np.exp(-x))
 
  def _relu_func(self, x):
    return np.maximum(0,x)
  
  def _tanh_func(self,x):
    return (2/(1+np.exp((-2)*x)))-1

  def fit(self,X,y):
    #samples and features 
    n_samples, n_features = X.shape

    #converting values to 1 and 0
    y_ = [1 if i>0 else 0 for i in y]

    #initializing weight and bias
    self.weight = np.zeros(n_features)
    self.bias = 0

    #gradient descenting
    for _ in range(self.n_iters):
      for idx, x_i in enumerate(X):
        lm = np.dot(x_i, self.weight) + self.bias
        y_prediction = self.activation_func(lm)

        #update rule
        update = self.lr*(y_[idx]- y_prediction)
        self.weight += update*x_i
        self.bias += update

  
  def predict(self,X):
    y_prediction = np.dot(X,self.weight) + self.bias
    return self.activation_func(y_prediction)


def accuracy(y_true, y_pred):
  return (np.sum(y_pred==y_true))/(len(y_true))


Above model creates a perceptron that can take different values for the learning rate, number of iterations and non-linear activation function such as


*   Step Function (step_function)
*   Signmoid Function (sigmoid)

*   RELU Function (relu)

*   Tanh Function (tanh)








#Credit Card Fraud Detection

In [17]:
!wget https://raw.githubusercontent.com/nsethi31/Kaggle-Data-Credit-Card-Fraud-Detection/master/creditcard.csv


--2021-05-03 06:05:54--  https://raw.githubusercontent.com/nsethi31/Kaggle-Data-Credit-Card-Fraud-Detection/master/creditcard.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 102634230 (98M) [text/plain]
Saving to: ‘creditcard.csv.1’

creditcard.csv.1    100%[===================>]  97.88M   161MB/s    in 0.6s    

2021-05-03 06:05:55 (161 MB/s) - ‘creditcard.csv.1’ saved [102634230/102634230]



In [18]:
import pandas as pd
from sklearn.model_selection import train_test_split


In [19]:
df = pd.read_csv("creditcard.csv")

In [20]:
df.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [21]:
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [23]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

There are no missing values in the dataset

In [24]:
df.Class.value_counts()

0    284315
1       492
Name: Class, dtype: int64

The given dataset is heavily imbalanced as seen from the counts, so undersampling method is used. 

After doing so, there will be same number of 0's and 1's in the dataset.

#Base Line Model

In [ ]:
df1 = df.copy()

In [ ]:
X = df1.drop("Class",axis =1)
y = df1.Class

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.3, random_state=42)

p = Perceptron(learning_rate=0.01, n_iters = 1000)


In [ ]:
X_train = np.array(X_train)
y_train = np.array(y_train)

In [ ]:
p.fit(X_train,y_train)

In [ ]:
y_pred = p.predict(X_test)
#print("learning rate: " + str(i) + " no. of iterations: " + str(j))
print(accuracy(y_test, y_pred))

0.9982093325374811


The baseline model gives an accuracy of 99.82%, which is invalid since the dataset is imbalanced. 



Models will be created using some undersampling methods

#Undersampling

In [92]:
X = df.drop("Class",axis=1)
y = df.Class

In [93]:
from imblearn.under_sampling import NearMiss
nm = NearMiss(random_state=42)
X_res,y_res=nm.fit_sample(X,y)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)


In [94]:
X_res.shape , y_res.shape

((984, 30), (984,))

In [95]:
X_train, X_test, y_train, y_test = train_test_split(X_res, y_res, test_size = 0.3, random_state=42)


In [96]:
#List of values of hyperparameters used while training the model

learning_rate_list = [0.01,0.001,0.0001]
n_iters_list = [10,100,1000]

##Step Function

In [97]:

accuracy_step_list=[]
learning_rate_list1 = []
n_iters_list1 = []
for i in learning_rate_list:
  for j in n_iters_list:
    p = Perceptron(i,j,function='step_function')
    p.fit(X_train,y_train)
    y_pred = p.predict(X_test)
    print("learning rate: " + str(i) + " no. of iterations: " + str(j))
    print(accuracy(y_test, y_pred)*100,'%')
    accuracy_step_list.append(accuracy(y_test, y_pred))
    learning_rate_list1.append(i)
    n_iters_list1.append(j)



learning rate: 0.01 no. of iterations: 10
51.35135135135135 %
learning rate: 0.01 no. of iterations: 100
53.71621621621622 %
learning rate: 0.01 no. of iterations: 1000
60.810810810810814 %
learning rate: 0.001 no. of iterations: 10
51.35135135135135 %
learning rate: 0.001 no. of iterations: 100
53.71621621621622 %
learning rate: 0.001 no. of iterations: 1000
60.810810810810814 %
learning rate: 0.0001 no. of iterations: 10
51.35135135135135 %
learning rate: 0.0001 no. of iterations: 100
53.71621621621622 %
learning rate: 0.0001 no. of iterations: 1000
60.810810810810814 %


In [98]:
def max_accuracy(accuracy_list, learning_rate_accuracy, no_iters_accuracy):
  max_accuracy = max(accuracy_list)
  i = accuracy_list.index(max_accuracy)
  lr = learning_rate_accuracy[i]
  ni = no_iters_accuracy[i]
  print("Non-linear function: ",p.activation_func)
  print("Maximum accuracy: ",(max_accuracy)*100,"%\nLearning rate: ",lr,"\nnumber of iterations: ",ni)

In [99]:
max_accuracy(accuracy_step_list,learning_rate_list1,n_iters_list1)

Non-linear function:  <bound method Perceptron._unit_step_func of <__main__.Perceptron object at 0x7f81218431d0>>
Maximum accuracy:  60.810810810810814 %
Learning rate:  0.01 
number of iterations:  1000


##Sigmoid Function

In [100]:
learning_rate_list2 = []
n_iters_list2 = []
accuracy_sigmoid_list=[]
for i in learning_rate_list:
  for j in n_iters_list:
    p = Perceptron(i,j,function='sigmoid')
    p.fit(X_train,y_train)
    y_pred = p.predict(X_test)
    print("learning rate: " + str(i) + " no. of iterations: " + str(j))
    print(accuracy(y_test, y_pred))
    accuracy_sigmoid_list.append(accuracy(y_test, y_pred))
    learning_rate_list2.append(i)
    n_iters_list2.append(j)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:22: RuntimeWarning: overflow encountered in exp


learning rate: 0.01 no. of iterations: 10
0.5135135135135135
learning rate: 0.01 no. of iterations: 100
0.7567567567567568
learning rate: 0.01 no. of iterations: 1000
0.668918918918919
learning rate: 0.001 no. of iterations: 10
0.5135135135135135
learning rate: 0.001 no. of iterations: 100
0.75
learning rate: 0.001 no. of iterations: 1000
0.5337837837837838
learning rate: 0.0001 no. of iterations: 10
0.5135135135135135
learning rate: 0.0001 no. of iterations: 100
0.5371621621621622
learning rate: 0.0001 no. of iterations: 1000
0.5337837837837838


In [101]:
max_accuracy(accuracy_sigmoid_list,learning_rate_list2,n_iters_list2)

Non-linear function:  <bound method Perceptron._sigmoid_func of <__main__.Perceptron object at 0x7f811e0811d0>>
Maximum accuracy:  75.67567567567568 %
Learning rate:  0.01 
number of iterations:  100


##Relu Function

In [102]:
learning_rate_list3 = []
n_iters_list3 = []
accuracy_relu_list=[]
for i in learning_rate_list:
  for j in n_iters_list:
    p = Perceptron(learning_rate = i,n_iters=j,function='relu')
    p.fit(X_train,y_train)
    y_pred = p.predict(X_test)
    print("learning rate: " + str(i) + " no. of iterations: " + str(j))
    print(accuracy(y_test, y_pred))
    accuracy_relu_list.append(accuracy(y_test, y_pred))
    learning_rate_list3.append(i)
    n_iters_list3.append(j)


learning rate: 0.01 no. of iterations: 10
0.5067567567567568
learning rate: 0.01 no. of iterations: 100
0.5067567567567568
learning rate: 0.01 no. of iterations: 1000
0.5067567567567568
learning rate: 0.001 no. of iterations: 10
0.5067567567567568
learning rate: 0.001 no. of iterations: 100
0.5067567567567568
learning rate: 0.001 no. of iterations: 1000
0.5067567567567568
learning rate: 0.0001 no. of iterations: 10
0.5067567567567568
learning rate: 0.0001 no. of iterations: 100
0.5067567567567568
learning rate: 0.0001 no. of iterations: 1000
0.5067567567567568


In [103]:
max_accuracy(accuracy_relu_list,learning_rate_list3,n_iters_list3)

Non-linear function:  <bound method Perceptron._relu_func of <__main__.Perceptron object at 0x7f812181a3d0>>
Maximum accuracy:  50.67567567567568 %
Learning rate:  0.01 
number of iterations:  10


##Tanh Function 

In [104]:
learning_rate_list4 = []
n_iters_list4 = []
accuracy_tanh_list=[]
for i in learning_rate_list:
  for j in n_iters_list:
    p = Perceptron(learning_rate = i,n_iters=j,function='tanh')
    p.fit(X_train,y_train)
    y_pred = p.predict(X_test)
    print("learning rate: " + str(i) + " no. of iterations: " + str(j))
    print(accuracy(y_test, y_pred))
    accuracy_tanh_list.append(accuracy(y_test, y_pred))
    learning_rate_list4.append(i)
    n_iters_list4.append(j)

learning rate: 0.01 no. of iterations: 10
0.0


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:28: RuntimeWarning: overflow encountered in exp


learning rate: 0.01 no. of iterations: 100
0.006756756756756757
learning rate: 0.01 no. of iterations: 1000
0.02702702702702703
learning rate: 0.001 no. of iterations: 10
0.0
learning rate: 0.001 no. of iterations: 100
0.006756756756756757
learning rate: 0.001 no. of iterations: 1000
0.02702702702702703
learning rate: 0.0001 no. of iterations: 10
0.0
learning rate: 0.0001 no. of iterations: 100
0.006756756756756757
learning rate: 0.0001 no. of iterations: 1000
0.02702702702702703


In [105]:
max_accuracy(accuracy_tanh_list,learning_rate_list4,n_iters_list4)

Non-linear function:  <bound method Perceptron._tanh_func of <__main__.Perceptron object at 0x7f811cfc0b90>>
Maximum accuracy:  2.7027027027027026 %
Learning rate:  0.01 
number of iterations:  1000
